File: CHUOND_4/4.19_FrameLayout_Java/app/src/main/java/com/example/myapplication/ui/CountryViewModel.java
package com.example.myapplication.ui;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.example.myapplication.data.Country;
import com.example.myapplication.data.CountryRepository;

import java.util.List;

public class CountryViewModel extends ViewModel {
    private final CountryRepository mCountryRepository;
    private final MutableLiveData<List<Country>> mCountries = new MutableLiveData<>();

    public CountryViewModel(CountryRepository countryRepository) {
        mCountryRepository = countryRepository;
    }

    public LiveData<List<Country>> getCountries() {
        return mCountries;
    }

    public void loadCountries() {
        List<Country> countries = mCountryRepository.loadCountries();
        mCountries.setValue(countries);
    }

    public static class Factory implements ViewModelProvider.Factory {
        private final CountryRepository mCountryRepository;

        public Factory(CountryRepository countryRepository) {
            mCountryRepository = countryRepository;
        }

        @SuppressWarnings("unchecked")
        @NonNull
        @Override
        public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
            if (modelClass.isAssignableFrom(CountryViewModel.class)) {
                return (T) new CountryViewModel(mCountryRepository);
            } else {
                throw new IllegalArgumentException("Unknown ViewModel class");
            }
        }
    }
}

File: CHUONG_10/10.2_ex1/app/src/main/java/com/example/myapplication/ui/CountryViewModel.kt
package com.example.myapplication.ui

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.example.myapplication.data.Country
import com.example.myapplication.data.LocalDataSource

class CountryViewModel (
    private val localDataSource: LocalDataSource
): ViewModel() {
    private val _countries = MutableLiveData<List<Country>>()
    val countries: LiveData<List<Country>> = _countries

    fun loadCountries() {
        _countries.value = localDataSource.getCountries()
    }

    class Factory(
        private val dataSource: LocalDataSource
    ) : ViewModelProvider.Factory {
        override fun <T : ViewModel> create(modelClass: Class<T>): T =
            when {
                modelClass.isAssignableFrom(CountryViewModel::class.java) ->
                    CountryViewModel(dataSource) as T
                else -> throw IllegalArgumentException("Unknown ViewModel class")
            }
    }
}